Simplify data structures, add uniform approximation, fix mem leak

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63164 Remove req to call step() in training loop
* #63163 Pass `_allow_empty_param_list` into func opt ctor
* **#63162 Simplify data structures, add uniform approximation, fix mem leak**

Differential Revision: [D30284617](https://our.internmc.facebook.com/intern/diff/D30284617)

**Overview:**
This adds an `bool` option `uniform_bucket_assignment` to `hook_with_zero_step()` and `hook_with_zero_step_interleaved()`. If `uniform_bucket_assignment=True`, then the overlapping approach tries to approximate a uniform bucket assignment using a heuristic. In particular, a capacity threshold is computed to be the total parameter size (computed by `Tensor.numel()`) divided by the world size, and a DDP bucket is partitioned subject to that capacity and greedily assigned to ranks. The DDP buckets are sorted by increasing size first. This targets the scenario when the world size exceeds the number of DDP buckets, resulting in several ranks without any shard of parameters to optimize.

This also performs some internal data structure cleanup in `zero_redundancy_optimizer.py`. Initially, I tried separating the cleanup into the first commit, leaving the new overlapping approach to the second commit, but it became difficult to isolate some of the changes.

Lastly, this fixes a memory leak resulting from the strong reference to `ddp` in the hooks.

**Test Plan:**
Correctness: I added a parity test for the two new approaches, and the accuracy benchmarks for each show parity as well.

**Miscellaneous Notes:**
The word "bucket" is becoming conflated since with this change, DDP buckets and ZeRO buckets are not necessarily in 1:1 correspondence. The language I preferred is to continue to refer to DDP buckets as "buckets", while referring to ZeRO buckets as "bucket assignments" (or more verbosely, a DDP bucket assignment). This captures that we are assigning a (possibly non-strict) subset of a DDP bucket to ZeRO.

Commit:
Simplify data structures, add uniform approximation, fix mem leak

[ghstack-poisoned]

andwgu commented:
@andwgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30284617).

mrshenli commented on torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:
```diff
@@ -64,58 +71,63 @@ def _broadcast_bucket(
+    assert len(overlap_info.assigned_ranks_per_bucket) > bucket_index, \
+        "`assigned_ranks_per_bucket` is not fully constructed"
+    # Sort to ensure the same ordering across ranks
+    assigned_ranks = sorted(overlap_info.assigned_ranks_per_bucket[bucket_index])
```
(no need to change anything) curious, do we need to sort this every iteration? I assume this list is static?

andwgu replied:
Currently, I have `overlap_info.assigned_ranks_per_bucket[bucket_index]` as a `set`, which is why I sort every time. We could alternatively make that a `list` and only sort once, but I do not think it makes a big difference either way.

mrshenli commented on torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:
```diff
@@ -64,58 +71,63 @@ def _broadcast_bucket(
+            overlap_info.broadcast_handles.append(
+                dist.broadcast(
+                    bucket_assignments[bucket_index].tensor,
+                    src=_get_global_rank(zero.process_group, assigned_rank),
```
When do we expect ZeRO's PG rank to be different from global rank? If this is not an expected case, shall we error out in the constructor instead?

andwgu replied:
https://github.com/pytorch/pytorch/issues/63213

mrshenli commented on torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:
```diff
@@ -64,58 +71,63 @@ def _broadcast_bucket(
+                dist.broadcast(
+                    bucket_assignments[bucket_index].tensor,
+                    src=_get_global_rank(zero.process_group, assigned_rank),
+                    group=zero.process_group,
```
hmm, we are using the ZeRO PG, but set the src rank to the global rank. If the local and global rank disagrees, e.g., when the global PG has a larger world size, this can crash?

andwgu replied:
https://github.com/pytorch/pytorch/issues/63213

mrshenli commented on torch/distributed/optim/zero_redundancy_optimizer.py:
```diff
@@ -158,6 +180,17 @@ class _OverlapInfo():
+        world_size (int): world size of the process group being used.
+
+    Attributes:
+        uniform_bucket_assignment (bool): if ``True``, then the assignment of
```
also mention that the threshold is `overlap_info.total_size / self.world_size`?

mrshenli replied:
will renaming this var to `shard_bucket` preserve the same meaning, but shorter?

mrshenli reviewed (approved):
LGMT. Left some minor comments.

Commit:
Update on "Simplify data structures, add uniform approximation, fix mem leak"


Differential Revision: [D30284617](https://our.internmc.facebook.com/intern/diff/D30284617)

**Overview:**
This adds an `bool` option `uniform_bucket_assignment` to `hook_with_zero_step()` and `hook_with_zero_step_interleaved()`. If `uniform_bucket_assignment=True`, then the overlapping approach tries to approximate a uniform bucket assignment using a heuristic. In particular, a capacity threshold is computed to be the total parameter size (computed by `Tensor.numel()`) divided by the world size, and a DDP bucket is partitioned subject to that capacity and greedily assigned to ranks. The DDP buckets are sorted by increasing size first. This targets the scenario when the world size exceeds the number of DDP buckets, resulting in several ranks without any shard of parameters to optimize.

This also performs some internal data structure cleanup in `zero_redundancy_optimizer.py`. Initially, I tried separating the cleanup into the first commit, leaving the new overlapping approach to the second commit, but it became difficult to isolate some of the changes.

Lastly, this fixes a memory leak resulting from the strong reference to `ddp` in the hooks.

**Test Plan:**
Correctness: I added a parity test for the two new approaches, and the accuracy benchmarks for each show parity as well.

**Miscellaneous Notes:**
The word "bucket" is becoming conflated since with this change, DDP buckets and ZeRO buckets are not necessarily in 1:1 correspondence. The language I preferred is to continue to refer to DDP buckets as "buckets", while referring to ZeRO buckets as "bucket assignments" (or more verbosely, a DDP bucket assignment). This captures that we are assigning a (possibly non-strict) subset of a DDP bucket to ZeRO.

[ghstack-poisoned]

andwgu commented:
@andwgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30284617).

andwgu commented on torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:
```diff
@@ -161,15 +179,20 @@ def hook_with_zero_step(
             "ZeroRedundancyOptimizer must be constructed with "
             "`overlap_with_ddp=True` to use this hook properly"
         )
+    ddp_ref = weakref.ref(ddp)
```
https://github.com/pytorch/pytorch/issues/63211
